src/Scheduler.js

```diff
@@ -1,55 +1,99 @@
-const Queue = require('./Queue'); 
-const { 
-    QueueType,
-    PRIORITY_LEVELS,
+const Queue = require('./Queue');
+const {
+  QueueType,
+  PRIORITY_LEVELS,
+  SchedulerInterrupt,
 } = require('./constants/index');
 
 // A class representing the scheduler
-// It holds a single blocking queue for blocking processes and three running queues 
+// It holds a single blocking queue for blocking processes and three running queues
 // for non-blocking processes
-class Scheduler { 
-    constructor() { 
-        this.clock = Date.now();
-        this.blockingQueue = new Queue(this, 50, 0, QueueType.BLOCKING_QUEUE);
-        this.runningQueues = [];
-        // Initialize all the CPU running queues
-        for (let i = 0; i < PRIORITY_LEVELS; i++) {
-            this.runningQueues[i] = new Queue(this, 10 + i * 20, i, QueueType.CPU_QUEUE);
-        }
+class Scheduler {
+  constructor() {
+    this.clock = Date.now();
+    this.blockingQueue = new Queue(this, 50, 0, QueueType.BLOCKING_QUEUE);
+    this.runningQueues = [];
+    // Initialize all the CPU running queues
+    for (let i = 0; i < PRIORITY_LEVELS; i++) {
+      this.runningQueues[i] = new Queue(
+        this,
+        10 + i * 20,
+        i,
+        QueueType.CPU_QUEUE
+      );
     }
+  }
 
-    // Executes the scheduler in an infinite loop as long as there are processes in any of the queues
-    // Calculate the time slice for the next iteration of the scheduler by subtracting the current
-    // time from the clock property. Don't forget to update the clock property afterwards.
-    // On every iteration of the scheduler, if the blocking queue is not empty, blocking work
-    // should be done. Once the blocking work has been done, perform some CPU work in the same iteration.
-    run() {
-
+  // Executes the scheduler in an infinite loop as long as there are processes in any of the queues
+  // Calculate the time slice for the next iteration of the scheduler by subtracting the current
+  // time from the clock property. Don't forget to update the clock property afterwards.
+  // On every iteration of the scheduler, if the blocking queue is not empty, blocking work
+  // should be done. Once the blocking work has been done, perform some CPU work in the same iteration.
+  run() {
+    while (!this.allQueuesEmpty()) {
+      console.log('WHILE LOOP');
+      const timeslice = Date.now() - this.clock;
+      if (!this.blockingQueue.isEmpty()) {
+        console.log('block empty');
+        this.blockingQueue.doBlockingWork(timeslice);
+      }
+      for (let i = 0; i < this.runningQueues.length; i++) {
+        console.log('FOR LOOP');
+        if (!this.runningQueues[i].isEmpty()) {
+          this.runningQueues[i].doCPUWork(timeslice);
+        }
+      }
     }
+    this.clock = Date.now();
+  }
 
-    allEmpty() {
-
+  allQueuesEmpty() {
```

Comment:
The tests aren't catching this (but that's going to be fixed), but this method should also check that the blocking queue is empty.